calendar_af/models/calendar_occasion.py

```diff
@@ -0,0 +1,728 @@
+                )
+            ):
+                occ_time[loop_date.strftime("%Y-%m-%dT%H:%M:%S")] = self.env[
+                    "calendar.occasion"
+                ].search_count(
+                    [("start", "=", loop_date), ("type_id", "=", type_id.id)]
+                )
+            loop_date = loop_date + timedelta(minutes=BASE_DURATION)
+        occ_time_min_key = min(occ_time, key=occ_time.get)
+        res = datetime.strptime(occ_time_min_key, "%Y-%m-%dT%H:%M:%S")
+        return res
+
+    @api.model
+    def _check_date_mapping(self, date, operation_id=False):
+        """Checks if a date has a mapped date, and returns the mapped date
+        if it exists"""
+        if operation_id:
+            mapped_date = self.env["calendar.mapped_dates"].search(
+                [("from_date", "=", date), ("operation_id", "=", operation_id.id)]
+            )
+        else:
+            mapped_date = self.env["calendar.mapped_dates"].search(
+                [("from_date", "=", date), ("operation_id", "=", False)]
+            )
+        if mapped_date:
+            res = mapped_date.to_date
+        else:
+            res = date
+        return res
+
+    @api.model
+    def _get_additional_booking(self, date, duration, type_id, operation_id=False):
+        """"Creates extra, additional, occasions. Iff overbooking is allowed. """
+        user_id = False
+        # Check if overbooking is allowed on this meeting type
+        if not type_id.additional_booking:
+            # TODO: Throw error instead?
+            _logger.debug(_("Overbooking not allowed on %s" % type_id.name))
+            return False
+        # Replace date with mapped date if we have one
+        date = self._check_date_mapping(date, operation_id)
+        date_list = date.strftime("%Y-%-m-%-d").split("-")
+        # Copy to make sure we dont overwrite BASE_DAY_START or BASE_DAY_STOP
+        day_start = copy(BASE_DAY_START)
+        day_stop = copy(BASE_DAY_STOP) - timedelta(hours=1)
+        # Ugly, ugly code..
+        day_start = day_start.replace(
+            year=int(date_list[0]), month=int(date_list[1]), day=int(date_list[2])
+        )
+        day_stop = day_stop.replace(
+            year=int(date_list[0]), month=int(date_list[1]), day=int(date_list[2])
+        )
+        res = self.env["calendar.occasion"]
+        # Find when to create new occasion
+        if (
+            type_id.channel == self.env.ref("calendar_channel.channel_local")
+            and operation_id
+            and operation_id.reserve_time
+        ):
+            date_now = datetime.now().replace(hour=0, minute=0, second=0, microsecond=0)
+            datetime_offset = timedelta(
+                days=type_id.days_last, hours=operation_id.reserve_time
+            )
+            start_date = date_now + datetime_offset
+            # Calculate how many occasions we need
+            no_occasions = int(duration / BASE_DURATION)
```

Comment:
Should this be round up, forcing it to a int rounds down... maybe is not a problem